src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -1413,86 +1389,68 @@ final ForkJoinTask<?> peek() {
 
         /**
          * Polls for a task. Used only by non-owners.
-         *
-         * @param pool if nonnull, pool to signal if more tasks exist
          */
-        final ForkJoinTask<?> poll(ForkJoinPool pool) {
-            for (;;) {
-                ForkJoinTask<?>[] a = array;
-                int b = base, cap, k;
-                if (a == null || (cap = a.length) <= 0)
-                    break;
-                ForkJoinTask<?> t = a[k = b & (cap - 1)];
-                U.loadFence();
-                if (base == b) {
-                    Object o;
-                    int nb = b + 1, nk = nb & (cap - 1);
-                    if (t == null)
-                        o = a[k];
-                    else if (t == (o = U.compareAndExchangeReference(
-                                       a, slotOffset(k), t, null))) {
+        final ForkJoinTask<?> poll() {
+            if ((phase & (IDLE | 1)) == 0 || top - base > 0) {
+                for (boolean stalled = false;;) {
+                    int cap, b, k, nb; ForkJoinTask<?>[] a;
+                    if ((a = array) == null || (cap = a.length) <= 0)
+                        break;
+                    long kp = slotOffset(k = (cap - 1) & (b = base));
+                    int nk = (nb = b + 1) & (cap - 1); // next slot
+                    int sk = (b + 2) & (cap - 1);      // 2nd slot ahead
+                    ForkJoinTask<?> t = a[k];
+                    U.loadFence();
+                    if (base != b)
+                        ;                              // inconsistent
+                    else if (t == null) {
+                        if (a[sk] == null && a[nk] == null && a[k] == null &&
+                            top - b <= 0)
+                            break;                     // empty
+                        if (stalled)
+                            Thread.onSpinWait();
+                        stalled = true;
```

Comment:
@DougLea Performance-wise, does it make any difference to make the write to `stalled` in an `else-branch` to avoid writing it every time beyond the initial?

Reply:
This doesn't matter performance-wise but I agree that it looks unnecessarily odd so will change.

Reply:
Or better (wrt this and previous comment): make the code-shape of poll() the same as inner loop of scan(). Done.